src/txn/txn_timestamp.c

```diff
@@ -658,318 +603,298 @@ __txn_assert_after_reads(
-		WT_RET_MSG(session, EINVAL, "setting a read_timestamp"
-		    " requires a transaction running at snapshot"
-		    " isolation");
-
-	/* Read timestamps can't change once set. */
-	if (F_ISSET(txn, WT_TXN_HAS_TS_READ))
-		WT_RET_MSG(session, EINVAL, "a read_timestamp"
-		    " may only be set once per transaction");
-
-	/*
-	 * This code is not using the timestamp validate function to
-	 * avoid a race between checking and setting transaction
-	 * timestamp.
-	 */
-	__wt_readlock(session, &txn_global->rwlock);
-	ts_oldest = txn_global->oldest_timestamp;
-	did_roundup_to_oldest = false;
-	if (read_ts < ts_oldest) {
-		/*
-		 * If given read timestamp is earlier than oldest
-		 * timestamp then round the read timestamp to
-		 * oldest timestamp.
-		 */
-		if (F_ISSET(txn, WT_TXN_TS_ROUND_READ)) {
-			txn->read_timestamp = ts_oldest;
-			did_roundup_to_oldest = true;
-		} else {
-			__wt_readunlock(session, &txn_global->rwlock);
-
-			/*
-			 * In some cases, MongoDB sets a read timestamp older
-			 * than the oldest timestamp, relying on WiredTiger's
-			 * concurrency to detect and fail the set. In other
-			 * cases it's a bug and MongoDB wants error context to
-			 * make it easier to find those problems. Don't output
-			 * an error message because that logs a MongoDB error,
-			 * use an informational message to provide the context
-			 * instead.
-			 */
-			WT_RET(__wt_msg(session, "read timestamp "
-			    "%s less than the oldest timestamp %s",
-			    __wt_timestamp_to_string(read_ts, ts_string[0]),
-			    __wt_timestamp_to_string(ts_oldest, ts_string[1])));
-			return (EINVAL);
-		}
-	} else
-		txn->read_timestamp = read_ts;
-
-	__wt_txn_publish_read_timestamp(session);
-	__wt_readunlock(session, &txn_global->rwlock);
-
-	/*
-	 * This message is generated here to reduce the span of critical
-	 * section.
-	 */
-	if (did_roundup_to_oldest)
-		__wt_verbose(session, WT_VERB_TIMESTAMP, "read "
-		    "timestamp %s : rounded to oldest timestamp %s",
-		    __wt_timestamp_to_string(read_ts, ts_string[0]),
-		    __wt_timestamp_to_string(ts_oldest, ts_string[1]));
-
-	/*
-	 * If we already have a snapshot, it may be too early to match
-	 * the timestamp (including the one we just read, if rounding
-	 * to oldest).  Get a new one.
-	 */
-	if (F_ISSET(txn, WT_TXN_RUNNING))
-		__wt_txn_get_snapshot(session);
-
-	return (0);
+    WT_TXN *txn = &session->txn;
+    WT_TXN_GLOBAL *txn_global = &S2C(session)->txn_global;
+    wt_timestamp_t ts_oldest;
+    char ts_string[2][WT_TS_INT_STRING_SIZE];
+    bool did_roundup_to_oldest;
+
+    WT_RET(__wt_txn_context_prepare_check(session));
+
+    /* Read timestamps imply / require snapshot isolation. */
+    if (!F_ISSET(txn, WT_TXN_RUNNING))
+        txn->isolation = WT_ISO_SNAPSHOT;
+    else if (txn->isolation != WT_ISO_SNAPSHOT)
+        WT_RET_MSG(session, EINVAL, "setting a read_timestamp"
+                                    " requires a transaction running at snapshot"
```

Comment:
String re-indents seem kinda weird?

Reply:
d7e9a08

Literals always get aligned like this. I made a change to always break before multi line strings so hopefully this looks less weird.